### ISSUE TITLE
defunct gpg-agents

### ISSUE THREAD

**Issue by NoUseFreak** [jkroepke/helm-secrets#200]
### Current Behavior

After a few minutes of running, the repo-server starts collecting defunct gpg-agents.

```
I have no name!@argocd-repo-server-54b5888987-2k2n4:/home/argocd$ helm plugin list
NAME   	VERSION	DESCRIPTION
secrets	3.12.0 	This plugin provides secrets values encryption for Helm charts secure storing
I have no name!@argocd-repo-server-54b5888987-2k2n4:/home/argocd$ ps aux
USER        PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
1000310+      1  0.2  0.2 788712 43076 ?        Ssl  10:16   0:00 argocd-repo-server --redis argocd-redis:6379 --logformat text --loglevel info
1000310+     16  0.0  0.0      0     0 ?        Z    10:16   0:00 [gpg-agent] <defunct>
1000310+     17  0.0  0.0  78272  1292 ?        Ss   10:16   0:00 gpg-agent --homedir /app/config/gpg/keys --use-standard-socket --daemon
1000310+    252  0.0  0.0      0     0 ?        Z    10:17   0:00 [gpg-agent] <defunct>
1000310+    255  0.0  0.0      0     0 ?        Zs   10:17   0:00 [gpg-agent] <defunct>
1000310+    280  0.0  0.0      0     0 ?        Z    10:17   0:00 [gpg-agent] <defunct>
1000310+    283  0.0  0.0      0     0 ?        Zs   10:17   0:00 [gpg-agent] <defunct>
1000310+    284  0.0  0.0      0     0 ?        Z    10:17   0:00 [gpg-agent] <defunct>
1000310+    292  0.0  0.0      0     0 ?        Zs   10:17   0:00 [gpg-agent] <defunct>
1000310+    449  0.0  0.0      0     0 ?        Z    10:17   0:00 [gpg-agent] <defunct>
1000310+    450  0.0  0.0      0     0 ?        Z    10:17   0:00 [gpg-agent] <defunct>
...
```

Eventually, the number of processes increases until the point where other applications are unable to spawn/fork processes.

### Expected Behavior

I've noticed a fix was implemented in https://github.com/jkroepke/helm-secrets/commit/4c31ab8f193d1f9b1ad5ff9412a0774f41ff4bb5 however, this does not seem to work.

### Steps To Reproduce

```markdown
Setup an ArgoCD application using sops + gpg.
Do a refresh in ArgoCD and run `ps aux` to see the number of remaining gpg-agents.
```

### Environment

- Helm Version: v3.6.0
- Helm Secrets Version: 3.12.0
- OS: ArgoCD image
- Shell: any

The ArgoCD installation was done using both custom image and downloader approaches. Same issue on both.

### Anything else?

If a workaround exists, for now, that would also be great!

**Comment by jkroepke** [comment]
Hi,

I can verify this, even after calling `gpgconf --kill gpg-agent` manually.


```
argocd@921dad2dff17:/tmp/tmp.5tCj5vNSVy/tmp.ZEz1eRoTcP/1OXIZm$ gpg -K
/tmp/tmp.5tCj5vNSVy/tmp.ZEz1eRoTcP/1OXIZm/pubring.kbx
-----------------------------------------------------
sec   rsa2048 2020-05-06 [SC]
      D6174A02027050E59C711075B430C4E58E2BBBA3
uid           [ unknown] jkroepke/helm-secrets DEMO (DONT USE THIS KEY IN PRODUCTION) <github@jkroepke.de>
ssb   rsa2048 2020-05-06 [E]

argocd@921dad2dff17:/tmp/tmp.5tCj5vNSVy/tmp.ZEz1eRoTcP/1OXIZm$ ps auxfw | grep 1249
argocd    1260  0.0  0.0   3392   840 pts/1    S+   00:04   0:00  \_ grep 1249
argocd    1249  0.0  0.1  78288  2684 ?        Ss   00:04   0:00 gpg-agent --homedir /tmp/tmp.5tCj5vNSVy/tmp.ZEz1eRoTcP/1OXIZm --use-standard-socket --daemon
argocd@921dad2dff17:/tmp/tmp.5tCj5vNSVy/tmp.ZEz1eRoTcP/1OXIZm$ ps auxfw | grep 1249^C
argocd@921dad2dff17:/tmp/tmp.5tCj5vNSVy/tmp.ZEz1eRoTcP/1OXIZm$ gpgconf --kill gpg-agent
argocd@921dad2dff17:/tmp/tmp.5tCj5vNSVy/tmp.ZEz1eRoTcP/1OXIZm$ ps auxfw | grep 1249
argocd    1265  0.0  0.0   3392   848 pts/1    S+   00:05   0:00  \_ grep 1249
argocd    1249  0.0  0.0      0     0 ?        Zs   00:04   0:00 [gpg-agent] <defunct>
```

A known solution would be [tini](https://github.com/krallin/tini#why-tini).

> It protects you from software that accidentally creates zombie processes, which can (over time!) starve your entire system for PIDs (and make it unusable).

**Comment by jkroepke** [comment]
Hi @NoUseFreak 

As short-term solution, I extend the documentation to integrate `tini`.
As long-term solution, I opened an issue https://github.com/argoproj/argo-cd/issues/8689 to solve this issue upstream.

**Comment by jkroepke** [comment]
Hi,

it turns out, tini is already included inside the ArgoCD Docker image, but the helm charts are not going to call tini binary.

https://github.com/argoproj/argo-helm/pull/1158#issuecomment-1059808368

**Comment by jkroepke** [comment]
Hi,

since this is strictly an issue at ArgoCD Helm Charts (https://github.com/argoproj/argo-helm/pull/1159), i'm going to close this.

**Comment by NoUseFreak** [comment]
Hi @jkroepke,

Thanks for all the effort and the fix. I can indeed confirm the issue is resolved when using the chart @ 3.35.2.

Great job!
